### PR TITLE
Add an example for Google Cloud Functions

### LINKED ISSUE #1369: Support vercel and nextjs Req and Res formats
**Is your feature request related to a problem? Please describe.**
Feature request, like there a lot of providers that uses custom AWS functions that wraps basic stuff.
`req` `res` etc.

**Describe the solution you'd like**
Redefining `generateCallback` will make sure that get full response so people can use telegraf, make it rewritable or as an option on configuration when you working on AWS lambda env.
```js
export default function (
    hookPath,
    updateHandler
) {
    return async (
        req,
        res,
        next = () => {
            res.status(403);
        }
) => {
        if (req.method !== 'POST' || req.url !== hookPath) {
            return next()
        }
        let update = req.body;

        try {
            await updateHandler(update, res)
        } catch (err) {
            res.status(500);
            throw err
        }
    }
}
```

For example: 
```js
const bot = new Telegraf(token, {
    telegram: { webhookReply: true, webhookReplyGenerator: (req, res) => req.body  }
})
```

```ts
  webhookCallback(path = '/') {
   const fn = this.telegram.webhookReply && this.telegram.webhookReplyGenerator  ? this.telegram.webhookReplyGenerator : generateCallback;
    return fn(
      path,
      (update: tt.Update, res: http.ServerResponse) =>
        this.handleUpdate(update, res)
    )
  }
```

**Describe alternatives you've considered**
-
**Additional context**
-

### PR DESCRIPTION
**Pull request by jacobwaller**
# Description

## Type of change
Small change to include a new example

# How Has This Been Tested?
Ran linter, no functional changes

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: docs/examples/google-cloud-function.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { Telegraf } from 'telegraf'
+
+const { BOT_TOKEN, PROJECT_ID, FUNCTION_NAME, REGION } = process.env
+
+// eslint-disable-next-line
+const bot = new Telegraf(`${BOT_TOKEN}`, {
+  telegram: { webhookReply: true },
```
Comment (wojpawlik) [inline docs/examples/google-cloud-function.ts]: This line does nothing here, please remove.

File: docs/examples/google-cloud-function.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+})
+
+// eslint-disable-next-line
```
Comment (wojpawlik) [inline docs/examples/google-cloud-function.ts]: Why is this needed?
```suggestion
```

File: docs/examples/google-cloud-function.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+export const botFunction = async (req: any, res: any) => {
+  console.log(req.body)
+  try {
+    await bot.handleUpdate(req.body)
+    console.log('Success')
+    res.status(200).send('Success')
+  } catch (err) {
+    console.log('Failure')
+    console.log(err)
+    res.status(500).send('Something went wrong')
```
Comment (wojpawlik) [inline docs/examples/google-cloud-function.ts]: This means that Telegram will resend the update and probably cause the same error again, right? Why is that desirable?

**Comment by jacobwaller** [comment]
@wojpawlik Not sure why you marked this PR as a draft, it is finished

**Review by wojpawlik — CHANGES_REQUESTED** [review]

File: docs/examples/google-cloud-function.ts (hunk 2 above)
Comment (jacobwaller) [reply docs/examples/google-cloud-function.ts]: setWebhook returns a promise and the eslint rules are set to not allow unhandled promises

File: docs/examples/google-cloud-function.ts (hunk 3 above)
Comment (jacobwaller) [reply docs/examples/google-cloud-function.ts]: No, the catch is to handle calls to the cloud function that are not from telegram

File: docs/examples/google-cloud-function.ts (hunk 2 above)
Comment (EdJoPaTo) [reply docs/examples/google-cloud-function.ts]: you can explictly mark which rule this is disabling:

```suggestion
// eslint-disable-next-line @typescript-eslint/no-floating-promises
```

File: docs/examples/google-cloud-function.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import { Telegraf } from 'telegraf'
+
+const { BOT_TOKEN, PROJECT_ID, FUNCTION_NAME, REGION } = process.env
+
+// eslint-disable-next-line
+const bot = new Telegraf(`${BOT_TOKEN}`)
+
+// eslint-disable-next-line @typescript-eslint/no-floating-promises
+bot.telegram.setWebhook(
+  // eslint-disable-next-line
```
Comment (EdJoPaTo) [inline docs/examples/google-cloud-function.ts]: Please specify exactly which rule is disabled

File: docs/examples/google-cloud-function.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import { Telegraf } from 'telegraf'
+
+const { BOT_TOKEN, PROJECT_ID, FUNCTION_NAME, REGION } = process.env
+
+// eslint-disable-next-line
```
Comment (EdJoPaTo) [inline docs/examples/google-cloud-function.ts]: Please specify exactly which rule is disabled

**Review by EdJoPaTo — CHANGES_REQUESTED** [review]
eslint can be a lot smarter when it knows which rules are disabled and will warn you when the rule is not disabled (anymore) because the rule got better for example.

Just disabling something is guesswork afterwards. Also I would prefer examples not needing bodges as that shouldn't be exemplary.

File: docs/examples/google-cloud-function.ts [hunk 6]
```diff
@@ -2,12 +2,12 @@ import { Telegraf } from 'telegraf'
 
 const { BOT_TOKEN, PROJECT_ID, FUNCTION_NAME, REGION } = process.env
```
Comment (EdJoPaTo) [inline docs/examples/google-cloud-function.ts]: You should check if these variables are set rather than disabling eslint warnings.
```ts
if (!BOT_TOKEN) {
  throw new Error('BOT_TOKEN required');
}
```

Having this check the first warning will be gone. Checking the other 3 and the second warning will be gone.

Typescript is there to help us, not to disable its warnings :)

Comment (jacobwaller) [reply docs/examples/google-cloud-function.ts]: The solution in that case would end up being overly specific. If you don't provide a bot token, telegram already throws an error. And `PROJECT_ID`, `FUNCTION_NAME`, and `REGION` are provided as env variables by default in Google Cloud Functions

Comment (jacobwaller) [reply docs/examples/google-cloud-function.ts]: Yes, I could check every env variable for undefined, but at that point, this example starts to become a little too complicated to be used as an example no?

Comment (EdJoPaTo) [reply docs/examples/google-cloud-function.ts]: If you are sure they are there maybe just use the `!` then.

```ts
const bot = new Telegraf(BOT_TOKEN!)
```

Seems more descriptive to me than ignoring eslint warnings.

File: docs/examples/google-cloud-function.ts (hunk 3 above)
Comment (wojpawlik) [reply docs/examples/google-cloud-function.ts]: How do you expect this to work? I'd expect you to check request path instead.

Please don't mark a conversation as resolved that I unresolved.
Also, I'll probably be inactive here for a while, focusing on 4.1.0 release. This will ship in 4.1.1.

Comment (jacobwaller) [reply docs/examples/google-cloud-function.ts]: I can just remove the error checking if the way I'm handling it isn't okay. Most of the other examples here don't have any

Comment (wojpawlik) [reply docs/examples/google-cloud-function.ts]: I'm unfamiliar with the platorm, but I'd do something akin to

```ts
try {
  await bot.handleUpdate(req.body)
finally {
  res.status(200)
}
```

Then if an update causes the bot to error, it won't be redelivered and won't cause another error.

Comment (jacobwaller) [reply docs/examples/google-cloud-function.ts]: Updated

File: docs/examples/google-cloud-function.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import { Telegraf } from 'telegraf'
+
+const BOT_TOKEN = process.env.BOT_TOKEN ?? '<BOT TOKEN HERE>'
+const PROJECT_ID = process.env.PROJECT_ID ?? '<Google Project ID>'
+const FUNCTION_NAME = process.env.FUNCTION_NAME ?? 'botFunction'
+const REGION = process.env.REGION ?? 'us-central1'
```
Comment (wojpawlik) [inline docs/examples/google-cloud-function.ts]: If these are always defined by Google, please just use non-null asserion

File: docs/examples/google-cloud-function.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+bot.command('hello', (ctx) => ctx.reply('Hello, friend!'))
+
+export const botFunction = async (req: any, res: any) => {
+  console.log(req.body)
+  try {
+    await bot.handleUpdate(req.body)
+    console.log('Success')
+  } catch (err) {
+    console.log(err)
```
Comment (wojpawlik) [inline docs/examples/google-cloud-function.ts]: `telegraf` alsready has (optional) logging
```suggestion
```

File: docs/examples/google-cloud-function.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import { Telegraf } from 'telegraf'
+
+const BOT_TOKEN = process.env.BOT_TOKEN ?? '<BOT TOKEN HERE>'
```
Comment (wojpawlik) [inline docs/examples/google-cloud-function.ts]: Please use empty string as default value or throw if it's `undefined`.

**Review by wojpawlik — CHANGES_REQUESTED** [review]
Please add this file to the `<ul>` in https://github.com/telegraf/telegraf/blob/develop/README.md#webhooks.

**Comment by jacobwaller** [comment]
For a project with very little documentation and examples, it sure is difficult to submit an example...

**Comment by wojpawlik** [comment]
Ok, I'll fix it myself

**Comment by wojpawlik** [comment]
@jacobwaller please verify I haven't messed anything up :)

**Review by wojpawlik — APPROVED** [review]

**Comment by jacobwaller** [comment]
> @jacobwaller please verify I haven't messed anything up :)

LGTM - Thanks!

File: docs/examples/google-cloud-function.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
+import { Telegraf } from 'telegraf'
+
+const { BOT_TOKEN, FUNCTION_NAME, PROJECT_ID, REGION } = process.env
+
+if (BOT_TOKEN === undefined) {
```
Comment (EdJoPaTo) [inline docs/examples/google-cloud-function.ts:7]: Maybe even check for falsy = empty strings?
```suggestion
if (!BOT_TOKEN) {
```

File: docs/examples/google-cloud-function.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+const bot = new Telegraf(BOT_TOKEN)
+
+// eslint-disable-next-line @typescript-eslint/no-floating-promises
+bot.telegram.setWebhook(
+  `https://${REGION!}-${PROJECT_ID!}.cloudfunctions.net/${FUNCTION_NAME!}`
+)
+
+bot.command('hello', (ctx) => ctx.reply('Hello, friend!'))
+
+export const botFunction = async (req: any, res: any) => {
```
Comment (EdJoPaTo) [inline docs/examples/google-cloud-function.ts]: I'm not sure with the platform. I assume this exported function is then used by the google cloud thingy?

Two questions here:

- Can we provide typings for the arguments? If this example is copied over users might find these typings helpful. But they are probably not as simple and would require an additional import?

- I would expect functions to be declared like functions and not like variables? Or is there any reason for this I am missing here? Or just personal preference?
```suggestion
export async function botFunction(req: any, res: any) {
```

**Review by EdJoPaTo — APPROVED** [review]

File: docs/examples/google-cloud-function.ts (hunk 10 above)
Comment (wojpawlik) [reply docs/examples/google-cloud-function.ts:7]: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/strict-boolean-expressions.md

Comment (EdJoPaTo) [reply docs/examples/google-cloud-function.ts:7]: Might be helpful as we dont have to know about this falsy stuff then.

Comment (wojpawlik) [reply docs/examples/google-cloud-function.ts:7]: Telegraf does the "sloppy" check internally anyway:

https://github.com/telegraf/telegraf/blob/66a0020e8522c81d600f2c309800aad058cacbd5/src/core/network/client.ts#L336-L341

It's a leftover from JS days, possibly still useful to people writing JS.

File: docs/examples/google-cloud-function.ts (hunk 11 above)
Comment (wojpawlik) [reply docs/examples/google-cloud-function.ts]: Here function syntax is purely a preference, not a problem, so I left it as-is.

More accurate parameter types would require a `devDependency` on `@types/express`.

Comment (EdJoPaTo) [reply docs/examples/google-cloud-function.ts]: Maybe just add a comment for this requirement then? Seems better than `any` to me. (I really don't like seeing `any`, especially in code meant to be copied.)
```ts
// You have to `npm install --save-dev @types/express` to use this example
import {Request, Response} from 'express'
```

But I assume the CI will be annoying about unknown types then…?

**Review by EdJoPaTo — APPROVED** [review]
